clvm/object_cache.py

```diff
@@ -0,0 +1,106 @@
+from typing import Callable, Dict, Generic, Optional, Tuple, TypeVar
+
+import hashlib
+
+from .CLVMObject import CLVMStorage
+
+T = TypeVar("T")
+
+
+class ObjectCache(Generic[T]):
+    """
+    `ObjectCache` provides a way to calculate and cache values for each node
+    in a clvm object tree. It can be used to calculate the sha256 tree hash
+    for an object and save the hash for all the child objects for building
+    usage tables, for example.
+
+    It also allows a function that's defined recursively on a clvm tree to
+    have a non-recursive implementation (as it keeps a stack of uncached
+    objects locally).
+    """
+
+    def __init__(self, f: Callable[["ObjectCache[T]", CLVMStorage], Optional[T]]):
+        """
+        `f`: Callable[ObjectCache, CLVMObject] -> Union[None, T]
+
+        The function `f` is expected to calculate its T value recursively based
+        on the T values for the left and right child for a pair. For an atom, the
+        function f must calculate the T value directly.
+
+        If a pair is passed and one of the children does not have its T value cached
+        in `ObjectCache` yet, return `None` and f will be called with each child in turn.
+        Don't recurse in f; that's part of the point of this function.
+        """
+        self.f = f
+        self.lookup: Dict[int, Tuple[T, CLVMStorage]] = dict()
+
+    def get(self, obj: CLVMStorage) -> T:
+        obj_id = id(obj)
+        if obj_id not in self.lookup:
```

Comment:
technically, I don't think this condition isn't necessary.  it's repeated on line 44.

Reply:
True enough. The counter is `.get` will be called frequently and this short-cut will prevent the list from being created in the more common "already cached" case.